src/compiler/transformers/module/module.ts

```diff
@@ -523,26 +519,65 @@ namespace ts {
                     return visitEndOfDeclarationMarker(<EndOfDeclarationMarker>node);
 
                 default:
-                    return visitEachChild(node, moduleExpressionElementVisitor, context);
+                    return nestedElementVisitor(node);
             }
         }
 
-        function moduleExpressionElementVisitor(node: Expression): VisitResult<Expression> {
-            // This visitor does not need to descend into the tree if there is no dynamic import or destructuring assignment,
-            // as export/import statements are only transformed at the top level of a file.
-            if (!(node.transformFlags & TransformFlags.ContainsDynamicImport) && !(node.transformFlags & TransformFlags.ContainsDestructuringAssignment)) {
-                return node;
-            }
+        function nestedElementVisitor(node: Node): Node {
+            switch (node.kind) {
+                case SyntaxKind.ForInStatement:
+                case SyntaxKind.ForOfStatement:
+                    return visitForInOrOfStatement(<ForInOrOfStatement>node);
+
+                case SyntaxKind.BinaryExpression:
+                    return visitBinaryExpression(<BinaryExpression>node, /*destructuringNeedsValue*/ true);
+
+                case SyntaxKind.PrefixUnaryExpression:
+                case SyntaxKind.PostfixUnaryExpression:
+                    return visitUnaryExpression(<PrefixUnaryExpression | PostfixUnaryExpression>node);
+
+                case SyntaxKind.ExpressionStatement:
+                    return visitExpressionStatement(<ExpressionStatement>node);
 
-            if (isImportCall(node)) {
-                return visitImportCallExpression(node);
+                default:
+                    if (isImportCall(node)) {
+                        return visitImportCallExpression(node);
+                    }
+                    return visitEachChild(node, nestedElementVisitor, context);

```

Comment:
Previously, `moduleExpressionElementVisitor` would only visit subtrees that contained either dynamic import or destructuring, meaning we didn't need to traverse the entire tree (which improves performance). This change results in a full walk of the entire tree. 

You can emulate the full walk during substitution by using emit notification to keep track of state as we descend through the tree during the final emit phase (which is always a full walk of the tree). You can see an example of this in many of the transformers, such as the `es2015` transformer which keeps track of the current state of the tree as we print so that substitutions can be aware of their context.

Reply:
Maybe I'm missing something, but as far as I understand this is not possible to implement using substitution and emit notification. That's the whole reason I rewrote the existing substitutions as visitor.
Some of the transformations need to create temporary variables. This seems to only be possible during the visiting stage. In addition I don't really want to reimplement the whole parenthesizing logic using emit notifications.